src/examples/jlang.md

```diff
@@ -0,0 +1,418 @@
+
+```pest
+COMMENT = _{ "NB." ~ ( !("\n" | EOI) ~ ANY)* }
+```
+
+## Parsing and AST Generation
+
+This section will walk through a parser that uses the pest grammar above. Library includes and self-explanatory code are omitted here; you can find the parser in its entirety at `examples/jlang-parser/src/main.rs` of this book's repository. 
+
+First we'll enumerate the verbs defined in our grammar, distinguishing between monadic and dyadic verbs. These enumerations will be be used as labels in our AST:
+
+```rust
+pub enum MonadicVerb {
+    Increment = 1,
+    Square = 2,
+    Negate = 3,
+    Reciprocal = 4,
+    Tally = 5,
+    Ceiling = 6,
+    ShapeOf = 7,
+}
+
+pub enum DyadicVerb {
+    Plus = 1,
+    Times = 2,
+    LessThan = 3,
+    LargerThan = 4,
+    Equal = 5,
+    Minus = 6,
+    Divide = 7,
+    Power = 8,
+    Residue = 9,
+    Copy = 10,
+    LargerOf = 11,
+    LargerOrEqual = 12,
+    Shape = 13,
+}
+```
+
+Then we'll enumerate the various kinds of AST nodes:
+
+```rust
+pub enum AstNode {
+    Print(Box<AstNode>),
+    Integer(i32),
+    DoublePrecisionFloat(f64),
+    MonadicOp { verb: MonadicVerb, expr: Box<AstNode> },
+    DyadicOp { verb: DyadicVerb, lhs: Box<AstNode>, rhs: Box<AstNode>},
+    Terms(Vec<AstNode>),
+    Reduce { verb: DyadicVerb, expr: Box<AstNode> },
+    IsGlobal{ident: String, expr: Box<AstNode>},
+    Ident(String),
+    Str(CString),
+}
+```
+
+To parse top-level statements in a J program, we have the following `parse` function that accepts a J program in string form and passes it to pest for parsing. We get back a sequence of pairs from pest. As specified in the grammar, a statement can only consist of an expression, so the `match` below parses each of those top-level expressions and wraps them in a `Print` AST node in keeping with the J interpreter's REPL behavior:
+
+```rust
+pub fn parse(source: &str) -> Result<Vec<AstNode>, Error<Rule>> {
+    let mut ast = vec![];
+    let pairs = JParser::parse(Rule::program, source)?;
+    for pair in pairs {
+        match pair.as_rule() {
+            Rule::expr => {
+                ast.push(Print(Box::new(build_ast_from_expr(pair))));
+            }
+            _ => {}
+        }
+    }
+    Ok(ast)
+}
+```
+
+AST nodes are built from expressions by walking pest's pair iterator in lockstep with the expectations set out in our grammar file. Common behaviors are abstracted out into separate functions, such as `parse_monadic_action` and `parse_dyadic_action`, and pairs representing expressions themselves are passed in recursive calls to `build_ast_from_expr`:
+
+```rust
+fn build_ast_from_expr(pair: pest::iterators::Pair<Rule>) -> AstNode {
+
+    match pair.as_rule() {
+        Rule::expr => build_ast_from_expr(pair.into_inner().next().unwrap()),
+        Rule::monadicExpr => {
```

Comment:
Some of these examples are very code heavy. Maybe it would make sense to have only one match pattern and the others omitted. Or maybe the second could point to an ellipsis (`=> { ... }`) and the others be omitted.